query-algebrizer/src/clauses/mod.rs

```diff
@@ -355,6 +334,73 @@ impl ConjoiningClauses {
         }
     }
 
+    /// Like `constrain_var_to_type` but in reverse: this expands the set of types
+    /// with which a variable is associated.
+    fn broaden_types(&mut self, additional_types: BTreeMap<Variable, HashSet<ValueType>>) {
+        for (var, new_types) in additional_types {
+            match self.known_types.entry(var) {
+                Entry::Vacant(e) => {
+                    if new_types.len() == 1 {
+                        self.extracted_types.remove(e.key());
+                    }
+                    e.insert(new_types);
+                },
+                Entry::Occupied(mut e) => {
+                    e.get_mut().extend(new_types.into_iter());
+                },
+            }
+        }
+    }
+
+    fn narrow_types_for_var(&mut self, var: Variable, types: HashSet<ValueType>) {
```

Comment:
I pushed a commit concurrent with your review that alters `extracted_types` a little.

We used to maintain `extracted_types` alongside `known_types`: if a var's type (only one!) was known, we didn't need a tag extraction column. If we later found out the type of an untyped var, then we'd remove the extraction column.

Now, of course, types can narrow and broaden. I killed most of the removals, and replaced them with a phase at the end.

The map still needs to exist, because it tracks which columns supply the tag.